Adds Rock Package Readme (#537)
ApollosProject/apollos-prototype: packages/apollos-data-connector-rock/README.md

# `@apollosproject/data-connector-rock`

Provides a connector to Rock for the following parts of the app

- Auth
- ContentItems
- ContentChannels
- Family
- Followings
- Interactions
- People
- PersonalDevices
- RockConstants
- Shareable
- Template

## Dependants / Dependencies

> None, right now. This module is 100% standalone. However, several modules internally require other modules internally. For example,

Reviewer: ```suggestion
None, right now. This module is 100% standalone. However, modules within the `data-connector` might depend on each other. For example,
```

- `People` and `Family` modules rely on the `Auth` module.
- `Followings` and `Interactions` relies on `RockConstants`.

## Usage

In your `src/data/index`,

```
import {
  Followings,
  Interactions,
  RockConstants,
  Family,
  Person,
  ContentItem,
  ContentChannel,
  Sharable,
  Auth,
  PersonalDevice,
  Template,
} from '@apollosproject/data-connector-rock';
import * as Theme from './theme';

...

const data = {
  ...
  Followings,
  Interactions,
  RockConstants,
  Family,
  Person,
  ContentItem,
  ContentChannel,
  Sharable,
  Auth,
  PersonalDevice,
  Template,

>   ...

Reviewer: Should we document the need to specify the `ContentItem` sub-types?

```
DevotionalContentItem: {
  dataSource: ContentItem,
}
```

I don't like that we have to do this..... fwiw this is required due to [this line](https://github.com/ApollosProject/apollos-prototype/blob/master/packages/apollos-server-core/src/node/model.js#L38). We expect there to be a `dataSource` present for each `type`. I've thought about a couple of times modifying the logic in `parseGlobalId` to look for a `dataSource` matching the `__type`'s `interface`... so for example if `DevotionalContentItem` didn't have it's own `dataSource` specified, it'd know to fall back to `ContentItem`. But I think that can get messy too :/

Author: Good catch 👌  I left a comment about ideas moving forward below.

  ...
}
```

Now your schema will have access to your new `contentItemsWithGroups` query. Congrats!

### Changing the behavior of existing fields.

Sometimes you may want to change the behavior of existing fields. For this example, we will override the firstName field of person so it always displays as all capital letters.

#### 1. Create a new module.
As before, this first thing you should do is create a new module, create a file/folder called `apollos-person/index.js`

```
import gql from graphql-tag;
import { People } from '@apollosproject/data-connector-rock';

const resolver = {};
```

#### 2. Define your overide.

We have a helper to enable safely overriding. You can import if from `@apollos/server-core`.

```
import gql from graphql-tag;
import { People } from '@apollosproject/data-connector-rock';
import { schemaMerge } from '@apollosproject/server-core';

const resolver = schemaMerge({
  Person: {
    firstName: ({ firstName }) => firstName.toUpperCase()
  }
}, People);

// make sure to export your resolver as resolver.

export { resolver };
```

#### 3. Import your new module.

Naming this time isn't important because we don't have a data source.

```

> // The import name is _very_ important. That will be the name of your dataSource.

Reviewer: This statement "the import name is _very_ important" seems to contradict with the line above "Naming this time isn't important". Am I reading this wrong?

Author: I made a bad copy/paste here 😂 I can delete this comment

import * as ApollosPerson from './apollos-person'

const data = {
  ...
  // make sure you put your override after the original person module.
  ApollosPerson,
}
```

#### Adding a new ContentItemType

Adding a new ContentItemType is a common operation, and we have some configuration tools to make it easier. Once you define your new ContentItem type's schema by extending the `ContentItem` GraphQL interface, add a new line to your `config.yml` in the RockMapppings -> ContentItem section. For example, if we knew that all of our GroupContentItem's referenced above had a content channel id of 27, we could add this to our config.yml in the RockMappings section.

```
    GroupContentItem:
      EntityType: ContentChannelItem
      ContentChannelTypeId: [27]
```

> If determining if the ContentItem type is more complex than something like ContentChannelTypeId, you'll need to override the `__resolveType` function on the ContentItem resolver.

Reviewer: ```suggestion
If determining the `ContentItem` type is more complex than something like` ContentChannelTypeId`, you'll need to override the `__resolveType` function on the `ContentItem` resolver.
```

Reviewer: Also, same as my comment above about linking the additional `dataSource`s for the new content item type - 

```
GroupContentItem: {
  dataSource: ContentItem,
}
```

Would this be needed in this case as well?
